crates/matrix-sdk/src/client/mod.rs

```diff
@@ -2243,8 +2243,15 @@ impl Client {
 
     /// Knock on a room given its `room_id_or_alias` to ask for permission to
     /// join it.
-    pub async fn knock(&self, room_id_or_alias: OwnedRoomOrAliasId) -> Result<Room> {
-        let request = knock_room::v3::Request::new(room_id_or_alias);
+    pub async fn knock(
+        &self,
+        room_id_or_alias: OwnedRoomOrAliasId,
+        reason: Option<String>,
+        server_names: Vec<OwnedServerName>,
+    ) -> Result<Room> {
+        let mut request = knock_room::v3::Request::new(room_id_or_alias);
+        request.reason = reason;
+        request.via = server_names;
```

Comment:
light suggestion: prefer using `assign!`:

```suggestion
        let request = assign!(knock_room::v3::Request::new(room_id_or_alias), { reason, via: server_names });
```